### ISSUE TITLE
Topbar icons and text are controlled by wrong variable

### ISSUE THREAD

**Issue by nickvergessen** [nextcloud/server#42954]
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are interested into the same feature.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

### Steps

* Use default background image
* Change theming color to the yellow option

Default blue | Option yellow
---|---
✅  | 💥  Text and all icons on the right side have the wrong color (based on primary and not if background is bright/dark, but that fact is used by the app icons in top left with `var(--background-image-invert-if-bright)` vs. `var(--color-primary-text)`
![Bildschirmfoto vom 2024-01-19 09-24-58](https://github.com/nextcloud/server/assets/213943/10e0eec3-c837-455e-8513-8c306cfb7ae7) | ![Bildschirmfoto vom 2024-01-19 09-24-50](https://github.com/nextcloud/server/assets/213943/456caa5a-e089-44bf-9917-703c09c1c926)

Default blue | Option yellow
---|---
💥  "More apps" icon has wrong color | 💥  "More apps" icon has wrong color 
![Bildschirmfoto vom 2024-01-19 09-25-12](https://github.com/nextcloud/server/assets/213943/f513a7b6-08e5-45d3-ae58-1008c0719312) | ![Bildschirmfoto vom 2024-01-19 09-25-22](https://github.com/nextcloud/server/assets/213943/dbb1bc70-6436-40b0-bd31-558243eb135e)

**Comment by JuliaKirschenheuter** [comment]
cc @szaimen

**Comment by JuliaKirschenheuter** [comment]
probably https://github.com/nextcloud/server/issues/42620 is a problem

**Comment by SystemKeeper** [comment]
<img width="349" alt="Bildschirmfoto 2024-01-19 um 10 40 16" src="https://github.com/nextcloud/server/assets/1580193/0aa43f23-c198-4857-bd61-a9180c220f1e">
<img width="351" alt="Bildschirmfoto 2024-01-19 um 10 39 51" src="https://github.com/nextcloud/server/assets/1580193/aecb7c17-0a82-4c19-b9af-0cc93464eb3b">

Might be related? Since the update to 28.0.2 RC1 theming returns black for text color for dark design and default background on c.nc.c.

**Comment by nickvergessen** [comment]
> Might be related? Since the update to 28.0.2 RC1 theming returns black for text color for dark design and default background on c.nc.c.

That's because the primary element color is now toned when contrast is too small.
But should be a different thing.

**Comment by szaimen** [comment]
> > Might be related? Since the update to 28.0.2 RC1 theming returns black for text color for dark design and default background on c.nc.c.
> 
> That's because the primary element color is now toned when contrast is too small. But should be a different thing.

Yes, this is unrelated

**Comment by susnux** [comment]
I think to make the color background image dependent it would work to use:

```css
color: white;
filter: var(--background-image-invert-if-bright);
```

Again this will not cover all cases especially for background images with shapes in different color.

**Comment by szaimen** [comment]
> I think to make the color background image dependent it would work to use:
> 
> ```css
> color: white;
> filter: var(--background-image-invert-if-bright);
> ```
> 
> Again this will not cover all cases especially for background images with shapes in different color.

yes, that is why I pointed at https://github.com/nextcloud/server/issues/35482#issuecomment-1331757066

**Comment by susnux** [comment]
Yes either each icon needs individual colors, or the background need a backdrop (see the comment below), or we simply ignore images that are not single color in the top area.

**Comment by ShGKme** [comment]
but we actually do have a color associated with each pre-defined background. Which can be used for defining elements on a background like the app menu.

https://github.com/nextcloud/server/blob/e5b0941c6446aeabeb40d3e98cd5e6e4cae2d07e/apps/theming/lib/Service/BackgroundService.php#L57-L71

Then it is used in `getColorPrimary`.

Cannot we then have separated background color and primary (aka elements) color?

**Comment by szaimen** [comment]
> Cannot we then have separated background color and primary (aka elements) color?

The problem is rather with custom primary colors. This also is taken into account for the color of the header icons which it should probably not be. However custom backgrounds are still broken because you cannot control the color of the header icons manually currently...

**Comment by ShGKme** [comment]
> The problem is rather with custom primary colors. This also is taken into account for the color of the header icons which it should probably not be. However custom backgrounds are still broken because you cannot control the color of the header icons manually currently...

Yes, but at least we may make app icons and text rely only on the background color, not on primary elements color.

**Comment by szaimen** [comment]
> Yes, but at least we may make app icons and text rely only on the background color, not on primary elements color.

Indeed that sounds better

**Comment by susnux** [comment]
I think we finally need to escape the primary-color-mixup-background-color rabbit hole.

The text color on the app menu should be independent from any primary color. Primary is only for accents, while the background color might sometimes be related but it is not always.

**Comment by ShGKme** [comment]
> The text color on the app menu should be independent from any primary color. Primary is only for accents, while the background color might sometimes be related but it is not always.

We may rename it or create a new variable, but anyway we have a color that represents the background color and store it together with the background image in presets...

**Comment by susnux** [comment]
Meaning:

* `primary-element` is what the user chose but made accessible
* `primary` is a bad name but should be the color to use on the background (-image, -color)
  * maybe rename to `color-background-text` or similar

**Comment by ShGKme** [comment]
> escape the primary-color-mixup-background-color rabbit hole

Having a good and full description of all the colors, variables, meanings we have and should have would be definitely helpful 😀

Taking into account user settings, custom colors, default color (Nextcloud Blue) and etc.

**Comment by susnux** [comment]
Maybe I find some time to look into it, but I think a proper solution contains real separation between background-color and primary-color.

Because:

> but we actually do have a color associated with each pre-defined background.

This color is just a primary color that probably looks good on that background but e.g. the bright image has some dark yellow set as primary which then causes primary-text to be white -> white text on white image -> bad!

**Comment by nursoda** [comment]
Since milestone 28.0.2 is mentioned above: It's not fixed in 28.0.2 – counting on 28.0.3.

**Comment by nursoda** [comment]
Please also fix this in /settings/apps while at it:
<img src="https://github.com/nextcloud/server/assets/8425483/f8fb27f0-4630-4cc0-a143-309765a4d78f" alt="Dashboard uses wrong icon in /settings/apps" width="314">

**Comment by susnux** [comment]
@nursoda this is unrelated - happens because of the icon SVG of the dashboard app.

**Comment by nursoda** [comment]
OK, so I assume this already is addressed in another issue.

**Comment by ObiWahn** [comment]
![screenshot_2024-02-17_11:02:11](https://github.com/nextcloud/server/assets/741109/8dad8b21-56e6-4602-8807-bee56f5e13f4)

please fix this, I have to deal with a lot of non technical users and they seem to care much more about this than about any other functional bug.

**Comment by escoand** [comment]
Duplicate of #43298

**Comment by susnux** [comment]
No this one is a different issue, this one is about the wrong color of the app bar icons when you select a different primary color.

In the example using the default clouds background image + yellow primary color will set black text of the app menu entries -> wrong color for the background image.

**Comment by susnux** [comment]
@ObiWahn looks fine on the screenshot, but you have set up a too bright background color which then causes the text color to switch to black.
The text color will always try to be at least 4.5:1 color contrast to the background color. This is needed for accessibility.

Try to set the background color to `#00679e` which is the default, this will then use white text and icons on the app navigation.

**Comment by ObiWahn** [comment]
Thanks @susnux, everything should be stock, I do not even know how to modify the themes.

After some testing:
- I have currently no background at all selected `index.php/settings/user/theming`
- And the current color is in the first col of the third row (afaik it was never modified).
- Switching to the 2nd col has the wanted effect
- I can not find any input textbox for a color code

Question:
- Is there a way make the slightly darker blue the default for all users?

**Comment by susnux** [comment]
> Is there a way make the slightly darker blue the default for all users?

https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/theming.html#theming

> And the current color is in the first col of the third row (afaik it was never modified).

Yes we needed to adjust the colors for accessibility and it seems we need to adjust the color picker defaults.
Once https://github.com/nextcloud/server/pull/42977 is merged this will be finally fixed.
